sycl/source/device_selector.cpp

```diff
@@ -28,8 +29,51 @@ static bool isDeviceOfPreferredSyclBe(const device &Device) {
          backend::level0;
 }
 
-device device_selector::select_device() const {
-  vector_class<device> devices = device::get_devices();
+// return a device with the requested deviceType (and requested backend)
+// if no such device is found, heuristic is used to select a device.
+// 'deviceType' is the desired device type
+//   info::device_type::all means it relies on the heuristic to select a device
+// 'be' is a specific desired GPU backend choice when multiple backends are
+// found.
+device device_selector::select_device(info::device_type deviceType,
+                                      backend be) const {
+  // return if a requested deviceType is found
+  if (deviceType != info::device_type::all) {
+    if (deviceType == info::device_type::host) {
+      vector_class<device> devices;
+      devices.resize(1);
+      return devices[0];
+    }
+
+    const vector_class<detail::plugin> &plugins = RT::initialize();
+    for (unsigned int i = 0; i < plugins.size(); i++) {
+      pi_uint32 numPlatforms = 0;
+      plugins[i].call<detail::PiApiKind::piPlatformsGet>(0, nullptr,
+                                                         &numPlatforms);
+      if (numPlatforms) {
```

Comment:
It seems you return nothing if numPlatforms == 0

Reply:
Yes, for the particular plugin[i] has no platforms, it won't return any device of course.
And then, it checks for other plugins[i+1], etc.
If all designated plugins do not find the requested device_type, then it goes to the legacy heuristic approach to find the best match (highest score).
The user will get a WARNING message to inform that select_device() chose a different device than your requested device_type.